server/api_archive_integration_test.go

```diff
@@ -0,0 +1,275 @@
+				t.Fatal(fmt.Errorf("file %v data does not match original file", zipf.Name))
+			}
+			zbuf := make([]byte, 64)
+			if !bytes.Equal(data[32:], zbuf) {
+				t.Fatal(fmt.Errorf("file %v zip contained private data contents", zipf.Name))
+			}
+		} else {
+			if !bytes.Equal(data, dataMap[zipf.Name]) {
+				t.Fatal(fmt.Errorf("file %v data does not match original file", zipf.Name))
+			}
+		}
+	}
+
+	for f, found := range fileMap {
+		if !found {
+			t.Fatal(fmt.Errorf("expected %v not found in zipfile", f))
+		}
+	}
+}
+
+func TestApiArchiveRateLimiterMultithreaded(t *testing.T) {
+	// Create a populated test environment and start a new server.
+	gcas, _, err := ServerTestEnvironment(t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer gcas.Close()
+
+	time.Sleep(apiArchiveRate) // Ensure that the rate limiter has cleared
+
+	var allowed atomic.Int32
+	var denied atomic.Int32
+	var wg sync.WaitGroup
+
+	// Choose a multiple of the api duration
+	maxDur := time.Duration(3) * apiArchiveRate
+
+	// Ensure the API is not called at the end of the duration
+	segment := maxDur / time.Duration(apiArchiveLimit+1)
+
+	start := time.Now()
+
+	const threads = 10
+
+	for i := 0; i < threads; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			// Randomly choose a point within each segment
+			durs := make([]float64, apiArchiveLimit)
+			for j := 0; j < apiArchiveLimit; j++ {
+				durs[j] = (float64(j) + rand.Float64()) * segment.Seconds()
+			}
+
+			slices.Sort(durs) // Make sure the times are in increasing order
+
+			cuttab := make([]time.Time, apiArchiveLimit)
+			for j := 0; j < apiArchiveLimit; j++ {
+				cuttab[j] = start.Add(time.Duration(durs[j] * float64(time.Second)))
+			}
+
+			for j := 0; j < apiArchiveLimit; j++ {
+				// Sleep until the next cut time.
+				time.Sleep(time.Until(cuttab[j]))
+
+				resp, err := http.Get(fmt.Sprintf("http://localhost:%v/api/v1/archive", gcas.httpPort))
+				if err != nil {
+					t.Fatal(err)
+				}
+				switch resp.StatusCode {
+				case http.StatusOK:
+					allowed.Add(1)
+				case http.StatusTooManyRequests:
+					denied.Add(1)
+				default:
+					t.Fatalf("Archive API call returned an invalid response %v", resp)
+				}
+			}
+		}()
+	}
+	wg.Wait()
+
+	testDur := time.Since(start)
+
+	if testDur > time.Duration(3)*apiArchiveRate {
+		t.Errorf("duration %v exceeded %v", testDur, time.Duration(3)*apiArchiveRate)
+	}
+
+	if allowed.Load()+denied.Load() != int32(threads*apiArchiveLimit) {
+		t.Errorf("incorrect requests %v, expected %v", allowed.Load()+denied.Load(), threads*apiArchiveLimit)
+	}
+
+	// Calculate the number of rate limit intervals.
+	periods := int(testDur / apiArchiveRate)
+	if testDur%apiArchiveRate != 0 {
+		periods++
+	}
+
+	if int(allowed.Load()) > periods*apiArchiveLimit {
+		t.Errorf("%v allowed, expected %v", allowed.Load(), periods*apiArchiveLimit)
+	}
+}
+
+// TODO: To avoid cut and paste, should consolidate some of these helper routines.
```

Comment:
The general rule of thumb in golang is to cut-and-paste until you've got more than three copies of a piece of code. The saying is "a little duplication is better than a little abstraction". The one exception would be if changes elsewhere in the code would cause problems in this test, but as long as everything is self contained it's often fine to just copy-paste